bin/i18n/utils/sync_up_base.rb

```diff
@@ -0,0 +1,93 @@
+require_relative '../i18n_script_utils'
+require_relative '../metrics'
+require_relative '../utils/crowdin_client'
+
+module I18n
+  module Utils
+    class SyncUpBase
+      Config = Struct.new :crowdin_project, :source_paths, :ignore_paths, keyword_init: true
+
+      MAIN_CROWDIN_PROJECT = 'codeorg'.freeze
+
+      def self.config
+        @config ||= Config.new(
+          crowdin_project: MAIN_CROWDIN_PROJECT,
+          source_paths: [],
+          ignore_paths: [],
+        )
+      end
+
+      def self.perform
+        i18n_resource_name = name[/^.*::(\w+::\w+)::SyncUp/, 1] || name
+
+        I18n::Metrics.report_runtime(i18n_resource_name, 'sync-up') do
+          new.send(:perform)
+        end
+      end
+
+      protected
+
+      def perform
+        progress_bar.start
+
+        crowdin_client # Initialize `@crowdin_client` outside threads to avoid multiple instance creations
+
+        I18nScriptUtils.process_in_threads(source_files, in_threads: PARALLEL_PROCESSES) do |source_file_path|
+          crowdin_client.upload_source_file(source_file_path)
+        ensure
+          mutex.synchronize {progress_bar.increment}
+        end
+
+        progress_bar.finish
+      end
+
+      private
+
+      PROGRESS_BAR_FORMAT = '%t: |%W| %c/%C %a'.freeze
+      PARALLEL_PROCESSES = 20 # https://developer.crowdin.com/api/v2/#section/Introduction/Rate-Limits
+
+      def config
+        self.class.config
+      end
+
+      def testing?
+        CDO.rack_env?(:development)
+      end
+
+      def crowdin_project
+        # uses test Crowdin projects that mirroring regular set of projects for testing purposes
```

Comment:
tiny grammar nitpick:
```suggestion
        # uses test Crowdin projects that mirror regular set of projects for testing purposes
```

Reply:
Alternatively, we could go with something slightly more verbose:

```suggestion
        # When testing, use a set of test Crowdin projects that mirrors our regular set of projects.
```

Whatever you think is more readable!